### PR TITLE
Change `is_step_change` to `change_annotations`

### LINKED ISSUE #580: Consider renaming `is_step_change`
>I want to say that 'step change' is not a term I am used to. Where is that an established term (happy to learn)? I have worked a lot with step functions in continuous space, but with discrete data points all changes look like a 'step change' :-) :shrug:. I think the term 'step change' points to (2) more than (1), but I'd hope we can find a nicer/more intuitive user-facing term, based on the consequences of the annotation. Based on what I've seen, I'd say: this is about being able to annotate change, significant change, expected to be permanent. Certainly reviewed change, hopefully (but not necessarily) explainable change -- with the goal to reset aspects about statistical analysis.

_Originally posted by @jgehrcke in https://github.com/conbench/conbench/issues/575#issuecomment-1375491063_

### LINKED ISSUE #581: Add JSON column to BenchmarkResult for "why" the result was a step change or outlier
>If the expected change is permanent, it could be lovely to enrich the annotation with a free text field where users should explain the reason for the change. This can be powerful also when improving performance over time: get an auto-generated plot that shows how performance improved over time and also _why_.

_Originally posted by @jgehrcke in https://github.com/conbench/conbench/issues/575#issuecomment-1375491063_

### PR DESCRIPTION
**Pull request by austin3dickey**
Fixes #581 and #580. 
Helps with #472, #473, #474, and #475.

This PR introduces `change_annotations`, a new free-form JSONB column on BenchmarkResult that is intended to store (post-analysis) information about how that BenchmarkResult stands among its distribution of results. In the future, it will store whether that result is the beginning of a step change/discontinuity/outlier/etc, as well as reasons for why it was annotated that way (user-given or bot-given).

This new functionality supersedes the `is_step_change` column so I removed that.

### CODE REVIEW

File: conbench/tests/api/_expected_docs.py
Comment (jonkeane) [inline conbench/tests/api/_expected_docs.py:1103]: We should add details about why one might set this field (and we should also be explicit that — for now — all details inside of it are entirely ignored). The prior boolean had "That is, when evaluating whether future results are regressions or improvements, should we treat data from before this result as incomparable?" which is IMO a decent description + vague enough for us to get to the end of this project without needing to come back and change it but still be meaningful to folks using conbench to have some idea of what to do with this.

**Review by jonkeane — COMMENTED** [review]
The code here looks ok to me and I'm fine to merge this and move on if it's helpful — but I'm a bit worried that this is more code than is necessary for us to get to our end goals for the history project so that we can start validating + exploring what kinds of discontinuities folks report + how to deal with those differently (if we even need to). 

The # of lines isn't huge, but there's a non-trivial maintenance burden for exposing more complications like this further up the stack (e.g. in #575 we'll need to validate json or add in some hardcoded json etc.) IMHO, we would be better served to leave this be as a boolean for now and get the history improvements in front of folks rather than try to figure out what the whole set of ways we might see deviations in the future + plan out + document + validate all of those (which could be connected to the spirit of as-much-as-needed-as-little-as-possible. 

But I won't block this from merging if you want to, let's do it (decisively) now and move on to the meat of this project.

File: conbench/tests/api/_expected_docs.py
Comment (austin3dickey) [reply conbench/tests/api/_expected_docs.py:1103]: Agreed - I will do this in #575 when I write code that needs a specific element of the JSON.

**Comment by austin3dickey** [comment]
>I'm a bit worried that this is more code than is necessary

My current hypothesis is that documenting and using the JSON field `change_annotations.begins_permanent_change` will only be a little bit harder than documenting and using `is_step_change`, and allows more flexibility for change later (no pun intended). I think that benefit is worth it. If I can't do #575 "easily," I'll reject that hypothesis, so maybe I won't merge this PR until that PR is rebased onto this one.

File: conbench/entities/benchmark_result.py
Comment (alistaire47) [inline conbench/entities/benchmark_result.py]: Is it more useful to make it nullable or have a default of `{}`? Not sure which is simpler to query

**Comment by jonkeane** [comment]
> If I can't do https://github.com/conbench/conbench/pull/575 "easily," I'll reject that hypothesis, so maybe I won't merge this PR until that PR is rebased onto this one.

Love this!

**Review by alistaire47 — APPROVED** [review]

File: conbench/entities/benchmark_result.py
Comment (austin3dickey) [reply conbench/entities/benchmark_result.py]: I'm not sure -- I did some basic googling earlier today and couldn't quickly find an answer. To me it "feels like" the {} is better because my Python code can always use `result.change_annotations.get("asd")` instead of something like `(result.change_annotations or {}).get("asd")`

Comment (austin3dickey) [reply conbench/entities/benchmark_result.py]: After a lot of discussion with @ElenaHenderson , I made this nullable. It's just simpler to do the migration that way.

**Pull request merged**
